# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;














# 🧪 Proyecto A/B Testing – TripleTen

## Evaluación del sistema de recomendaciones en tienda en línea

👤 **Nombre:** Edna Sánchez  
📅 **Fecha de entrega:** Junio 2025

---

### 🎯 Objetivo del Proyecto

Evaluar el impacto de un nuevo sistema de recomendaciones (Grupo B) frente al sistema actual (Grupo A) en una tienda online de la región de la UE.

La métrica principal es la mejora de la conversión en tres eventos del embudo:

- `product_page`
- `product_cart`
- `purchase`

Se espera un aumento del **10% o más** en cada etapa para considerar exitoso el experimento.

---

### 📁 Datasets Utilizados

| Archivo                                  | Descripción |
|------------------------------------------|-------------|
| `final_ab_events_upd_us.csv`             | Eventos registrados de los usuarios (product_page, cart, purchase) |
| `final_ab_new_users_upd_us.csv`          | Nuevos usuarios registrados entre el 7 y 21 de diciembre |
| `final_ab_participants_upd_us.csv`       | Participantes asignados a los grupos A y B |
| `ab_project_marketing_events_us.csv`     | Campañas de marketing realizadas durante el experimento (solo contextual) |

---


## 📥 1. Importación de datos

Comenzamos cargando los archivos de datos necesarios para la prueba A/B. Validaremos su estructura para asegurar que todo esté correcto.

### Archivos utilizados:
- `ab_project_marketing_events_us.csv`: calendario de eventos de marketing.
- `final_ab_events_upd_us.csv`: eventos generados por usuarios.
- `final_ab_new_users_upd_us.csv`: usuarios nuevos registrados.
- `final_ab_participants_upd_us.csv`: asignación de usuarios a grupos de prueba.


In [1]:
import pandas as pd

# Intento de carga
try:
    marketing_events = pd.read_csv('/datasets/ab_project_marketing_events_us.csv')
except Exception as e:
    print("❌ Error cargando marketing_events:", e)

# Cargar los demás
events = pd.read_csv('/datasets/final_ab_events_upd_us.csv')
new_users = pd.read_csv('/datasets/final_ab_new_users_upd_us.csv')
participants = pd.read_csv('/datasets/final_ab_participants_upd_us.csv')

# Verificamos estructura básica
print("🗃️ Eventos:")
print(events.info(), '\n')

print("🧑 Nuevos usuarios:")
print(new_users.info(), '\n')

print("🧪 Participantes:")
print(participants.info(), '\n')


🗃️ Eventos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423761 entries, 0 to 423760
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   user_id     423761 non-null  object 
 1   event_dt    423761 non-null  object 
 2   event_name  423761 non-null  object 
 3   details     60314 non-null   float64
dtypes: float64(1), object(3)
memory usage: 12.9+ MB
None 

🧑 Nuevos usuarios:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58703 entries, 0 to 58702
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     58703 non-null  object
 1   first_date  58703 non-null  object
 2   region      58703 non-null  object
 3   device      58703 non-null  object
dtypes: object(4)
memory usage: 1.8+ MB
None 

🧪 Participantes:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14525 entries, 0 to 14524
Data columns (total 3 columns):
 #   Column   Non-Null Coun

## 🧹 2. Preprocesamiento de Datos

Revisamos:
- Tipos de datos y conversiones necesarias
- Valores nulos o faltantes
- Duplicados
- Rangos y consistencia en las fechas

Además, convertiremos los datos de fecha a formato datetime y revisaremos la validez del periodo de prueba.


In [2]:
# Copia de seguridad de los datos originales
events_raw = events.copy()
new_users_raw = new_users.copy()
participants_raw = participants.copy()

# Convertimos fechas a datetime
events['event_dt'] = pd.to_datetime(events['event_dt'])
new_users['first_date'] = pd.to_datetime(new_users['first_date'])

# Verificamos rangos de fechas
print("📆 Eventos desde:", events['event_dt'].min(), "hasta:", events['event_dt'].max())
print("📆 Registros de nuevos usuarios desde:", new_users['first_date'].min(), "hasta:", new_users['first_date'].max())

# Revisamos valores nulos
print("\n🔍 Valores nulos por dataset:")
print("Events:\n", events.isna().sum(), '\n')
print("New Users:\n", new_users.isna().sum(), '\n')
print("Participants:\n", participants.isna().sum(), '\n')

# Revisamos duplicados
print("📑 Duplicados:")
print("Events:", events.duplicated().sum())
print("New Users:", new_users.duplicated().sum())
print("Participants:", participants.duplicated().sum())


📆 Eventos desde: 2020-12-07 00:00:33 hasta: 2020-12-30 23:36:33
📆 Registros de nuevos usuarios desde: 2020-12-07 00:00:00 hasta: 2020-12-23 00:00:00

🔍 Valores nulos por dataset:
Events:
 user_id            0
event_dt           0
event_name         0
details       363447
dtype: int64 

New Users:
 user_id       0
first_date    0
region        0
device        0
dtype: int64 

Participants:
 user_id    0
group      0
ab_test    0
dtype: int64 

📑 Duplicados:
Events: 0
New Users: 0
Participants: 0


## 🔎 3. Validación del Rango de Fechas y Filtro de Usuarios

Vamos a:
- Confirmar que todos los usuarios en el grupo A/B se registraron dentro del periodo oficial del experimento.
- Asegurar que ningún usuario esté presente en ambos grupos.
- Filtrar únicamente a los usuarios relevantes para la prueba A/B.


In [3]:
# Periodo oficial del experimento
start_date = pd.to_datetime('2020-12-07')
cutoff_date = pd.to_datetime('2020-12-21')

# Filtramos usuarios registrados en el periodo de prueba
valid_users = new_users.query('first_date >= @start_date and first_date <= @cutoff_date')

# Filtramos participantes que realmente entraron en el experimento
participants_filtered = participants[participants['user_id'].isin(valid_users['user_id'])]

# Revisamos cantidad de usuarios por grupo
print("📊 Participantes válidos por grupo:")
print(participants_filtered['group'].value_counts())

# Verificamos si algún usuario está en ambos grupos (NO debería ocurrir)
duplicated_users = participants_filtered.duplicated(subset='user_id', keep=False)
users_in_both_groups = participants_filtered[duplicated_users]

print("\n❗ Usuarios presentes en ambos grupos (esperado: 0):", users_in_both_groups['user_id'].nunique())


📊 Participantes válidos por grupo:
A    7725
B    5798
Name: group, dtype: int64

❗ Usuarios presentes en ambos grupos (esperado: 0): 887


## 🔧 4. Limpieza de Participantes con Doble Asignación

Se detectaron **887 usuarios** asignados simultáneamente a ambos grupos A y B, lo cual rompe el supuesto de exclusividad en pruebas A/B. Procedemos a eliminarlos del análisis para mantener la validez estadística.


In [4]:
# Detectamos los user_id duplicados (que aparecen en más de un grupo)
duplicated_ids = participants_filtered['user_id'][participants_filtered.duplicated('user_id', keep=False)]

# Filtramos los participantes excluyendo los duplicados
clean_participants = participants_filtered[~participants_filtered['user_id'].isin(duplicated_ids)]

# Validamos los resultados
print("✅ Participantes por grupo (después de limpieza):")
print(clean_participants['group'].value_counts())

print("\n✅ Total usuarios excluidos por doble asignación:", len(duplicated_ids.unique()))


✅ Participantes por grupo (después de limpieza):
A    6604
B    5145
Name: group, dtype: int64

✅ Total usuarios excluidos por doble asignación: 887


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - Muy buena implementación para detectar y eliminar usuarios duplicados. Este paso es esencial para asegurar la independencia de los grupos. Claridad en los prints y seguimiento del conteo de participantes limpian adecuadamente el experimento.

</div>

## 🔗 5. Integración de Datos

Ahora que hemos limpiado los datos y eliminado a los usuarios con asignación duplicada, filtramos los eventos para quedarnos solo con aquellos realizados por usuarios válidos en el test. Luego, agregamos la información del grupo (A o B) a cada uno de estos eventos.


In [5]:
# Filtrar eventos para incluir solo los de usuarios que participaron correctamente en el experimento
valid_events = events[events['user_id'].isin(clean_participants['user_id'])]

# Unir información del grupo A o B a cada evento
merged_data = valid_events.merge(clean_participants[['user_id', 'group']], on='user_id', how='left')

# Verificamos que todo esté bien
print("📌 Eventos después de filtrado:", merged_data.shape[0])
print("👀 Eventos por grupo:")
print(merged_data['group'].value_counts())


📌 Eventos después de filtrado: 85902
👀 Eventos por grupo:
A    48625
B    37277
Name: group, dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - Integraste correctamente los datasets y filtraste solo los eventos de usuarios válidos. Esto garantiza que las comparaciones de métricas se basen en datos limpios y bien contextualizados. Muy buena estructura y legibilidad del código.

</div>

## 🧪 6. Análisis del Embudo de Conversión

Analizaremos las tasas de conversión de los usuarios en tres etapas clave:

- `product_page` → vista de producto
- `product_cart` → añadir al carrito
- `purchase` → compra

Comparamos las tasas de conversión por grupo (A vs B) para evaluar el impacto del nuevo sistema de recomendaciones.


In [6]:
# Lista de eventos clave
events_funnel = ['product_page', 'product_cart', 'purchase']

# Número de usuarios únicos por grupo y evento
funnel = merged_data[merged_data['event_name'].isin(events_funnel)] \
    .groupby(['group', 'event_name'])['user_id'].nunique() \
    .unstack().fillna(0)

# Agregar número total de usuarios por grupo
funnel['total_users'] = clean_participants.groupby('group')['user_id'].nunique()

# Calcular tasas de conversión
for event in events_funnel:
    funnel[f'{event}_conversion'] = funnel[event] / funnel['total_users']

# Redondear para visualizar mejor
funnel = funnel.round(4)

display(funnel)


event_name  product_cart  product_page  purchase  total_users  \
group                                                           
A                   2110          4391      2264         6604   
B                   1711          3311      1677         5145   

event_name  product_page_conversion  product_cart_conversion  \
group                                                          
A                            0.6649                   0.3195   
B                            0.6435                   0.3326   

event_name  purchase_conversion  
group                            
A                        0.3428  
B                        0.3259

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - Excelente diseño del embudo. Calculaste conversiones por etapa con claridad y aplicaste proporciones en función del total de usuarios, no de eventos, lo cual es correcto. Bien redondeado para facilitar la lectura. Buena práctica incluir todos los eventos relevantes.

</div>

## 📊 7. Prueba de Hipótesis – Z-Test para Proporciones

Realizaremos una prueba de hipótesis para determinar si las diferencias observadas en las tasas de conversión entre los grupos A y B son estadísticamente significativas.

- **H₀ (nula)**: No hay diferencia significativa entre los grupos.
- **H₁ (alternativa)**: Hay diferencia significativa en las conversiones.

Nivel de significancia: α = 0.05


In [7]:
from statsmodels.stats.proportion import proportions_ztest

# Función para hacer Z-test
def z_test(success_a, size_a, success_b, size_b):
    count = [success_a, success_b]
    nobs = [size_a, size_b]
    stat, pval = proportions_ztest(count, nobs)
    return round(stat, 4), round(pval, 4)

# Resultados de prueba por etapa
tests = {}

# product_page
tests['product_page'] = z_test(
    funnel.loc['A', 'product_page'],
    funnel.loc['A', 'total_users'],
    funnel.loc['B', 'product_page'],
    funnel.loc['B', 'total_users']
)

# product_cart
tests['product_cart'] = z_test(
    funnel.loc['A', 'product_cart'],
    funnel.loc['A', 'total_users'],
    funnel.loc['B', 'product_cart'],
    funnel.loc['B', 'total_users']
)

# purchase
tests['purchase'] = z_test(
    funnel.loc['A', 'purchase'],
    funnel.loc['A', 'total_users'],
    funnel.loc['B', 'purchase'],
    funnel.loc['B', 'total_users']
)

# Mostrar resultados
for etapa, (z, p) in tests.items():
    print(f'{etapa.upper()}: Z = {z}, p = {p}')


PRODUCT_PAGE: Z = 2.4176, p = 0.0156
PRODUCT_CART: Z = -1.4984, p = 0.134
PURCHASE: Z = 1.9221, p = 0.0546


## ✅ 8. Resultados del Test A/B – Z-Test para Proporciones

| Etapa         | Z-Score | Valor p  | ¿Significativo? (α = 0.05) |
|---------------|---------|----------|-----------------------------|
| product_page  | 2.4176  | 0.0156   | ✅ Sí                       |
| product_cart  | -1.4984 | 0.1340   | ❌ No                       |
| purchase      | 1.9221  | 0.0546   | ❌ No (casi significativo)  |

### 🧠 Interpretación

- Solo la diferencia en la etapa de **product_page** es estadísticamente significativa (p < 0.05), lo que sugiere que el nuevo sistema de recomendaciones puede estar afectando positivamente esta etapa del embudo.
- Las etapas de **product_cart** y **purchase** no muestran diferencias estadísticamente significativas, aunque la etapa de **purchase** está cerca del umbral.
- Ninguna de las diferencias alcanza el **10% de mejora esperada** planteada como criterio de éxito.


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Atención</b> ⚠️ - El análisis es correcto, pero podrías fortalecer la visualización de los resultados con una tabla más formal o gráficos de barras para comparar visualmente conversiones y significancia. También podrías enfatizar que no se alcanzó el umbral del 10% en ninguna etapa.

</div>

## ✅ 9. Validación de Supuestos Estadísticos

Antes de aplicar la prueba Z, se comprobaron los siguientes criterios:

- ✔️ **Tamaño muestral suficiente**: ambos grupos superan las 5,000 observaciones, lo cual garantiza aproximación normal por el Teorema Central del Límite.
- ✔️ **Independencia**: Se eliminaron usuarios duplicados (n=887) presentes en ambos grupos.
- ⚠️ **Homogeneidad de varianza**: Considerada indirectamente en el uso de proporción combinada para la prueba Z. No hay evidencia clara de sesgo.

🔍 Con base en estos puntos, aplicar una prueba Z es **metodológicamente válido** en este contexto.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - Excelente que valides explícitamente los supuestos del test estadístico. Esto es una práctica avanzada que muchos omiten. Además, mencionas correctamente el Teorema Central del Límite y los supuestos de independencia.

</div>

## ✅ 10. Conclusiones Finales del Proyecto

Realizamos un análisis de prueba A/B para evaluar si un nuevo sistema de recomendaciones mejoraba la conversión en las etapas `product_page → product_cart → purchase`.

### 🔍 Hallazgos clave:

- **product_page**: mejora significativa en visualizaciones de productos (Z = 2.42, p = 0.0156)
- **product_cart**: sin diferencia significativa (p = 0.134)
- **purchase**: casi significativa (p = 0.0546), pero no supera el umbral de 0.05 ni cumple con el +10% esperado.

### 📉 Conclusión general:

Aunque el sistema genera más visitas a productos, **no se traduce en mejoras en el embudo completo**. No se alcanzan los criterios definidos como éxito.

### ✅ Recomendaciones:

- 📈 Mejorar el algoritmo de recomendaciones con foco en conversión.
- 🧪 Realizar una nueva prueba A/B con mayor muestra, duración y segmentación (por región/dispositivo).
- 🧠 Complementar con análisis cualitativo para entender bloqueos en el embudo de conversión.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - Conclusión clara y bien argumentada. Resumes con precisión los hallazgos y limitaciones. Las recomendaciones son razonables y demuestran pensamiento crítico. Muy profesional que incluyas sugerencias prácticas para mejorar el experimento.

</div>

## 📚 11. Fuentes y Documentación

- Guía del proyecto final – TripleTen: [Notion Project Link](https://coding-bootcamps.notion.site/PROYECTO-FINAL-DATA-ANALYST-3fe741c1f13849d691b4afe136b68c72)
- Pandas Documentation: https://pandas.pydata.org/docs/
- Scipy Z-Test: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.proportions_ztest.html
- Matplotlib + Seaborn: Documentación oficial
- Stack Overflow – Resolución de errores


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - Buena inclusión de enlaces útiles, lo cual refuerza la transparencia y facilita la reproducibilidad del trabajo. Es una práctica profesional que siempre suma puntos en reportes formales.

</div>